### PR TITLE
[Finder] Check PHP version before applying a workaround for a PHP bug

### PR DESCRIPTION
**Pull request by alebo**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | -- |
| License | MIT |
| Doc PR | -- |

### CODE REVIEW

File: src/Symfony/Component/Finder/Iterator/FilterIterator.php [hunk 1]
```diff
@@ -12,9 +12,9 @@
 namespace Symfony\Component\Finder\Iterator;
 
 /**
- * This iterator just overrides the rewind method in order to correct a PHP bug.
+ * This iterator just overrides the rewind method in order to correct a PHP bug, which existed before version 5.5.23.
```
Comment (Tobion) [inline src/Symfony/Component/Finder/Iterator/FilterIterator.php]: https://bugs.php.net/bug.php?id=68557 says

> Confirmed fixed under PHP 5.5.22.

Comment (alebo) [inline src/Symfony/Component/Finder/Iterator/FilterIterator.php]: I double-checked before making the change. It's not fixed in 5.5.22. Also, the fix is listed only in
http://php.net/ChangeLog-5.php#5.5.23

File: src/Symfony/Component/Finder/Iterator/FilterIterator.php [hunk 2]
```diff
@@ -28,22 +29,24 @@
      */
     public function rewind()
     {
-        $iterator = $this;
-        while ($iterator instanceof \OuterIterator) {
-            $innerIterator = $iterator->getInnerIterator();
+        if (version_compare(PHP_VERSION, '5.5.23', '<')
```
Comment (xabbuh) [inline src/Symfony/Component/Finder/Iterator/FilterIterator.php]: We use the `PHP_VERSION_ID` constant for this:

```php
if (PHP_VERSION_ID < 50523 || PHP_VERSION_ID >= 50600 && PHP_VERSION_ID < 50607) {
    // ...
}
```

File: src/Symfony/Component/Finder/Iterator/FilterIterator.php [hunk 3]
```diff
@@ -28,22 +29,23 @@
      */
     public function rewind()
     {
-        $iterator = $this;
-        while ($iterator instanceof \OuterIterator) {
-            $innerIterator = $iterator->getInnerIterator();
+        if (PHP_VERSION_ID < 50523 || PHP_VERSION_ID >= 50600 && PHP_VERSION_ID < 50607) {
```
Comment (dunglas) [inline src/Symfony/Component/Finder/Iterator/FilterIterator.php]: You can use a guard close, it will improve the readability of this method:

```if (PHP_VERSION_ID > 50607 || (PHP_VERSION_ID > 50523 && PHP_VERSION_ID < 50600)) {
    parent:: rewind();
   return;
}
```

File: src/Symfony/Component/Finder/Iterator/FilterIterator.php [hunk 4]
```diff
@@ -12,9 +12,10 @@
 namespace Symfony\Component\Finder\Iterator;
 
 /**
- * This iterator just overrides the rewind method in order to correct a PHP bug.
+ * This iterator just overrides the rewind method in order to correct a PHP bug,
+ * which existed before version 5.5.23/5.6.7.
  *
- * @see https://bugs.php.net/bug.php?id=49104
+ * @see https://bugs.php.net/bug.php?id=68557
```
Comment (jakzal) [inline src/Symfony/Component/Finder/Iterator/FilterIterator.php]: Why did you change this?

Comment (alebo) [inline src/Symfony/Component/Finder/Iterator/FilterIterator.php]: @jakzal, it was me who added the original reference and at that time that PHP bug seemed to be related to the experienced problem. Now it turns out that the latter bug is the actual reason. The former one is still not fixed but it doesn't seem to cause any incorrect behavior.

Comment (jakzal) [inline src/Symfony/Component/Finder/Iterator/FilterIterator.php]: Cool. Thanks for the update!

File: src/Symfony/Component/Finder/Iterator/RecursiveDirectoryIterator.php [hunk 5]
```diff
@@ -118,8 +118,10 @@ public function rewind()
             return;
         }
 
-        // @see https://bugs.php.net/bug.php?id=49104
-        parent::next();
+        // @see https://bugs.php.net/bug.php?id=68557
+        if (PHP_VERSION_ID < 50523 || PHP_VERSION_ID >= 50600 && PHP_VERSION_ID < 50607) {
```
Comment (alebo) [inline src/Symfony/Component/Finder/Iterator/RecursiveDirectoryIterator.php]: @dunglas, Do you think this one also needs to be changed to a guard clause? I think that the same condition should be used in both cases, but here duplication of the parent::rewind call in the guard clause and in the normal flow would look weird.

File: src/Symfony/Component/Finder/Iterator/FilterIterator.php (hunk 4 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Finder/Iterator/FilterIterator.php]: btw, the link could be shortened as `https://bugs.php.net/68557` :)

File: src/Symfony/Component/Finder/Iterator/FilterIterator.php (hunk 3 above)
Comment (fabpot) [inline src/Symfony/Component/Finder/Iterator/FilterIterator.php]: :+1:

File: src/Symfony/Component/Finder/Iterator/RecursiveDirectoryIterator.php (hunk 5 above)
Comment (fabpot) [inline src/Symfony/Component/Finder/Iterator/RecursiveDirectoryIterator.php]: This one looks fine to me.

**Comment by fabpot** [comment]
Thank you @alebo.
